### DISCUSSION TITLE
Perpetual Looping of Autotuner?

### DISCUSSION THREAD

**Discussion by dalyles** [The-OpenROAD-Project/OpenROAD-flow-scripts#1138] [General]
I am trying to do a run on the <code>swerv</code> example circuit from the library <code>nangate45</code>.
What seems to be happening is that there is a supposed infite loop of running the iterations when
I run the command:

<code>python3 distributed.py --design swerv --platform nangate45  --config ../designs/nangate45/swerv/autotuner.json tune</code>

<img width="1638" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/b2a9ca6f-0e5a-436f-a428-baf81e0d7412">

Luckily, the contents in the <code>config.mk</code>, <code>constraint.sdc</code>, and  <code>autotuner.json</code> are small enough to display here.  If this helps I'm using python version 3.10.9.
All other files in <code>OpenROAD-flow-scripts/flow/designs/nangate45/swerv</code> are left unchanged.

<code>config.mk</code>
<img width="546" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/63a1f99b-db9e-4e45-a983-c84460144243">

<code>constraint.sdc</code>
<img width="604" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/1d82ab8f-e6f9-4321-a7cb-af116b85dbdd">

<code>autotuner.json</code>
<img width="302" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/c5be3c4a-f95d-43a4-ade2-78d3f7746616">

Thank you very much in advance.

**Comment by vijayank88** [comment]
@dalyles 
Are you running locally or jobs distributed?

**Reply by dalyles** [reply]
I'm running it on a remote server. I'm not sure if that completely answers your question.

**Reply by vijayank88** [reply]
You have to follow instruction from here: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/tools/AutoTuner/README.md to distribute jobs for complex designs. Otherwise it may run for log time.

**Reply by dalyles** [reply]
@vijayank88 

Do you have a tutorial or a walkthrough on how to run this with Autotuner?
Thanks

**Comment by dalyles** [comment]
Another thing I want to add is that it looks like during the supposed infinite loop, each iteration is exactly the same.

**Reply by vijayank88** [reply]
I've used following `autotuner.json` with Centos 7/128GB RAM/32 Cores cloud server. 
```
{
    "_SDC_FILE_PATH": "constraint.sdc",
    "_SDC_CLK_PERIOD": {
        "type": "float",
        "minmax": [
            1.0,
            3.4
        ],
        "step": 0
    },
    "CORE_UTILIZATION": {
        "type": "int",
        "minmax": [
            5,
            100
        ],
        "step": 1
    },
    "CORE_ASPECT_RATIO": {
        "type": "float",
        "minmax": [
            0.5,
            2.0
        ],
        "step": 0
    },
    "CORE_MARGIN": {
        "type": "int",
        "minmax": [
            2,
            2
        ],
        "step": 0
    },
    "CELL_PAD_IN_SITES_GLOBAL_PLACEMENT": {
        "type": "int",
        "minmax": [
            0,
            5
        ],
        "step": 1
    },
    "CELL_PAD_IN_SITES_DETAIL_PLACEMENT": {
        "type": "int",
        "minmax": [
            0,
            5
        ],
        "step": 1
    },
    "_FR_LAYER_ADJUST": {
        "type": "float",
        "minmax": [
            0.1,
            0.7
        ],
        "step": 0
    },
    "PLACE_DENSITY_LB_ADDON": {
        "type": "float",
        "minmax": [
            0.0,
            0.99
        ],
        "step": 0
    },
    "_PINS_DISTANCE": {
        "type": "int",
        "minmax": [
            1,
            4
        ],
        "step": 1
    },
    "CTS_CLUSTER_SIZE": {
        "type": "int",
        "minmax": [
            10,
            200
        ],
        "step": 1
    },
    "CTS_CLUSTER_DIAMETER": {
        "type": "int",
        "minmax": [
            20,
            400
        ],
        "step": 1
    },
    "_FR_FILE_PATH": "../../../platforms/nangate45/fastroute.tcl",
    "_FR_GR_OVERFLOW": {
        "type": "int",
        "minmax": [
            1,
            1
        ],
        "step": 0
    }
}
```
Detail routing going on for `variant-AutoTunerBase-accfae84-ray`. All other combinations got terminated. All happened in 3 Hrs duration.
![Screenshot from 2023-06-22 16-14-41](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/85787913/c49fe043-e710-428c-8c12-88a9ffb22bce)

**Reply by vijayank88** [reply]
Best parameters found:
```
2023-06-22 11:00:03,104	INFO tune.py:639 -- Total run time: 10852.13 seconds (10851.97 seconds for the tuning loop).
[INFO TUN-0002] Best parameters found: {'_SDC_CLK_PERIOD': 3.2433794787353936, 'CORE_UTILIZATION': 98, 'CORE_ASPECT_RATIO': 1.288325817455164, 'CORE_MARGIN': 2, 'CELL_PAD_IN_SITES_GLOBAL_PLACEMENT': 0, 'CELL_PAD_IN_SITES_DETAIL_PLACEMENT': 1, '_FR_LAYER_ADJUST': 0.39813722784548444, 'PLACE_DENSITY_LB_ADDON': 0.42029200119725857, '_PINS_DISTANCE': 1, 'CTS_CLUSTER_SIZE': 126, 'CTS_CLUSTER_DIAMETER': 313, '_FR_GR_OVERFLOW': 1}
(save_best pid=10002) [INFO TUN-0003] Best parameters written to /home/vijayan/OPENROAD_FLOW/ORFS_LOCAL/OpenROAD-flow-scripts/flow/logs/nangate45/swerv/test-tune-2023-06-22-07-59-06/autotuner-best-ac039f88.json
```

**Reply by vvbandeira** [reply]
@dalyles, what makes you think the iterations are the same? Note that the "Status" section is printed periodically and does not reflect an iteration. Note also that the run that finished, i.e., with the "TERMINATED" status, took over 4h to run. Finally, are you interested in the `swerv` design or only in testing the AutoTuner? If it is the latter, I would suggest you start with `gcd`, which has a much smaller runtime.

**Reply by dalyles** [reply]
Hi @vvbandeira,

For the Autotuner, I'm trying to test it, yes.
However I am using <code>swerv</code> since it has a variety of violations such as
maximum capacitance violations and maximum slew rate violations.
<code>gcd</code> only gives a setup time violation when running with default parameters.

For <code>swerv</code> I'm adding the parameters <code>CAP_MARGIN</code> and <code>SLEW_MARGIN</code> into the <code>autotuner.json</code> file and the <code>constraint.sdc</code> to see if will successfully converge and then run the flow using the best parameters found.

**Reply by vvbandeira** [reply]
@dalyles, thanks for the info. Was my explanation about the log files clear/useful? Could you expand a little on why you think the experiments are always the same?

**Reply by dalyles** [reply]
Hi @vvbandeira,
Yes your explanation was helpful. Just to confirm what you were saying, I reran the autotuner and let it run overnight and I saw that progress is being made although it is quite slow being slow. From what I understand, just because the console output periodically displays the status, doesn't mean it is redoing the same iteration. The same iteration is still in progress. I'll take your recommendation that you posted in your other comment.

**Comment by vijayank88** [comment]
@vvbandeira Trimming `autotuner.json` variable will cause any issue, because user mention it is running infinitely.

@dalyles can you share OS name/RAM/No.of cores in remote server?

**Reply by vvbandeira** [reply]
It should not cause any issues, but it can affect the runtime depending on the combination of parameters and how hard OR has to work (specially DRT).

**Reply by dalyles** [reply]
Here is the information you requested:

OS: Ubuntu 20.04.03 LTS
Memory: 512GB (16x 32GB DRR4 2667 MT/s)
CPU: 2x AMD EPYC 7453 28-Core Processor @ 2.75GHz (28 Core, 2 Thread, 64MB L3)

Hope this helps.

**Reply by vvbandeira** [reply]
That is a reasonable HW to run `swerv` AT in small batches experiments. A couple of settings to help you take better advantage of your HW. You can increase `resources_per_trial` to improve the runtime of each trial. `samples` and `iterations` will help improve the final result: one iteration is just parameter sweeping, and more iterations == more chance of getting a better result; one sample is too small, the larger sample size will improve results but increase total wall time.

## global

```
 --jobs <int>          Max number of concurrent jobs.
```

## `tune` specific

```
--samples <int>       Number of samples for tuning.
--iterations <int>    Number of iterations for tuning.
--resources_per_trial <int> Number of CPUs to request for each tunning job.
```

**Reply by dalyles** [reply]
Thank you very much for this, I'll give this a shot, so I just append these options in my original command?
Sorry I'm a bit of a novice when it comes to these things.

**Reply by dalyles** [reply]
So I tried using this command: 

<code>:~/OpenROAD-flow-scripts/flow/util$ python3 distributed.py  --design swerv --platform nangate45  --config ../designs/nangate45/swerv/autotuner.json tune --resources_per_trial 10 --jobs 10</code>

<code>usage: distributed.py [-h] --design <gcd,jpeg,ibex,aes,...> --platform <sky130hd,sky130hs,asap7,...> --config <path> [--experiment <str>] [--timeout <float>] [--git_clean] [--git_clone] [--git_clone_args <str>] [--git_latest] [--git_or_branch <str>]
                      [--git_orfs_branch <str>] [--git_url <url>] [--build_args <str>] [--jobs <int>] [--openroad_threads <int>] [--server <ip|servername>] [--port <int>] [-v]
                      {tune,sweep} ...</code>

<code>distributed.py: error: unrecognized arguments: --jobs 10</code>

It sees to not recognize the argument for <code>jobs</code> but it is fine with accepting <code>resources_per_trial</code>

**Reply by vvbandeira** [reply]
Global arguments should go before the `tune` or `sweep` keywords, i.e., the `--jobs` -- the other three can go in any order after the `tune` keyword.

---

I would suggest, based on your HW above, `--resources_per_trial 14 --samples 24 --iterations 3`. This will run 24 trials, in 3 batches of 8 -- since each trial uses 14 CPUs it can consume all 112 threads of your system (`14*8=112`) if the 8 parallel traials reach DRT at the same time. Expected runtime for this would be 3 * single run of swerv, in your case from the original post 4h, thus `4h*3` = 12h (overnight :) )

**Reply by dalyles** [reply]
But why would you want to have a longer runtime? Wouldn't it make more sense to utilize the resources to reduce the runtime?

**Reply by vvbandeira** [reply]
AutoTuner needs data to improve the results it provides. The higher the number of iterations, the higher the probability of better parameters. A higher number of samples per iteration means more data for AT to use in the next iteration. Choosing the "correct" amount of samples and iterations is an open research question. In short, if everything is well set up, the higher the runtime, the better parameters you will get -- note that this is a very generalized statement; there's an upper bound limit for the tool and the search space (i.e., `--config <FILE>.json`).
